### PR TITLE
rgw: Do not send a Content-Length header on status 204

### PR DESCRIPTION
**Pull request by wido**
RFC7230 says:
  A server MUST NOT send a Content-Length header field in any response with a
  status code of 1xx (Informational) or 204 (No Content).

Fixes: #13582

### CODE REVIEW

**Comment by wido** [comment]
@yehudasa Could you maybe take a look at this one? It's currently breaking running RGW behind a Varnish proxy.

**Comment by wido** [comment]
@yehudasa Thank you for the comments, see my new commit

File: src/rgw/rgw_civetweb.h [hunk 1]
```diff
@@ -19,6 +19,7 @@ class RGWMongoose : public RGWClientIO
   bufferlist data;
 
   int port;
+  int status_num;
```
Comment (yehudasa) [inline src/rgw/rgw_civetweb.h]: @wido status_num is missing initialization

File: src/rgw/rgw_fcgi.h [hunk 2]
```diff
@@ -13,6 +13,7 @@ struct FCGX_Request;
 class RGWFCGX : public RGWClientIO
 {
   FCGX_Request *fcgx;
+  int status_num;
```
Comment (yehudasa) [inline src/rgw/rgw_fcgi.h]: @wido this one too

**Comment by yehudasa** [comment]
@wido looking better, see my comments

File: src/rgw/rgw_civetweb.h (hunk 1 above)
Comment (wido) [inline src/rgw/rgw_civetweb.h]: Should it? Isn't a int always zero (0) in C++ when defined?

Comment (yehudasa) [inline src/rgw/rgw_civetweb.h]: @wido no, need to explicitly initialize it.

**Comment by wido** [comment]
@yehudasa Can you take another look? status_num should be initialized now.

**Comment by wido** [comment]
@yehudasa Ok, again :)

**Comment by wido** [comment]
ping @yehudasa

**Comment by yehudasa** [comment]
@wido see https://github.com/ceph/ceph/pull/6975

**Comment by wido** [comment]
Closing this one because #6975 seems better :)

**Pull request closed without merging**